roaring/roaring.go

```diff
@@ -3637,3 +3988,85 @@ func readWithRuns(b *Bitmap, data []byte, pos int, keyN uint32) {
 		}
 	}
 }
+
+// handledIter and handledIters are wrappers around Bitmap Container iterators
+// and assist with the unionIntoTarget algorithm by abstracting away some tedious
+// operations.
+type handledIter struct {
+	iter    ContainerIterator
+	hasNext bool
+	handled bool
+}
+
+type handledIters []handledIter
+
+func (w handledIters) next() bool {
+	hasNext := false
+
+	for i, wrapped := range w {
+		next := wrapped.iter.Next()
+		w[i].hasNext = next
+		w[i].handled = false
+		if next {
+			hasNext = true
+		}
+	}
+
+	return hasNext
+}
+
+func (w handledIters) markItersWithCurrentKeyAsHandled(key uint64) {
+	for i, wrapped := range w {
+		currKey, _ := wrapped.iter.Value()
+		if currKey == key {
+			w[i].handled = true
+		}
+	}
+}
+
+func (w handledIters) calculateSummaryStats(key uint64) containerUnionSummaryStats {
+	summary := containerUnionSummaryStats{}
+
+	for _, iter := range w {
+		if summary.hasMaxRange {
+			// If we already know that we're going to use a max range RLE container,
+			// then there is no reason to continue calculating statistics.
+			continue
+		}
+
+		// Calculate key-level statistics here
+		currKey, currContainer := iter.iter.Value()
+
+		if key == currKey {
+			summary.isOnlyContainerWithKey = false
+			summary.n += currContainer.n
+			if !summary.hasMaxRange {
+				summary.hasMaxRange = (currContainer.n == maxContainerVal+1)
```

Comment:
So far as I can tell, if this conditional ends up true, we are actually done. We could omit all tests in this function against summary.hasMaxRange, and just `if currContainer.n == maxContainerVal+1 { summary.hasMaxRange = true; return summary }`.